packages/navigator/src/browser/open-editors-widget/navigator-open-editors-file-decorator.ts

```diff
@@ -0,0 +1,112 @@
+/********************************************************************************
+ * Copyright (C) 2021 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from '@theia/core/shared/inversify';
+import { TreeDecorator, TreeDecoration } from '@theia/core/lib/browser/tree/tree-decorator';
+import { Emitter } from '@theia/core/lib/common/event';
+import { Tree } from '@theia/core/lib/browser/tree/tree';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { ApplicationShell, DepthFirstTreeIterator, LabelProvider, Saveable } from '@theia/core/lib/browser';
+import URI from '@theia/core/lib/common/uri';
+import { OpenEditorNode } from './navigator-open-editors-tree-model';
+import { EditorPreviewWidget } from '@theia/editor-preview/lib/browser';
+import { Disposable } from '@theia/core/lib/common';
+
+export interface OpenEditorTreeDecorationData extends TreeDecoration.Data {
```

Comment:
Would it work not to have this as a field on the decoration (which upsets TS a little bit), but evaluate the dirty status in the widget when rendering each node? Then you could fire an event here when the dirty status changed to trigger a rerender in the tree, but not modify the `TreeDecoration.Data` type. Alternatively, you could just move all the dirty-tracking logic to the `TreeModel` or the `Tree` and trigger a rerender in either of those places.

Reply:
Good thinking, I've made this change as well. Feels a good bit cleaner, and turns out I was already firing the event for dirty status change